aptos-move/framework/aptos-framework/sources/nonce_validation.move

```diff
@@ -0,0 +1,207 @@
+            };
+            move_to<NonceHistory>(aptos_framework, nonce_history);
+        };
+    }
+
+    fun empty_bucket(): Bucket {
+        let bucket = Bucket {
+            nonce_maps: vector[
+                Map {
+                    nonces: big_ordered_map::new_with_config(0, 0, true),
+                    last_stored_time: 0,
+                },
+                Map {
+                    nonces: big_ordered_map::new_with_config(0, 0, true),
+                    last_stored_time: 0,
+                },
+            ]
+        };
+        bucket.nonce_maps[0].nonces.allocate_spare_slots(5);
+        bucket.nonce_maps[1].nonces.allocate_spare_slots(5);
+        bucket
+    }
+
+    // This method is used to prefill the nonce_table with empty buckets one by one.
+    public entry fun add_nonce_buckets(count: u64) acquires NonceHistory {
+        assert!(exists<NonceHistory>(@aptos_framework), error::invalid_state(E_NONCE_HISTORY_DOES_NOT_EXIST));
+        let nonce_history = &mut NonceHistory[@aptos_framework];
+        for (i in 0..count) {
+            if (nonce_history.next_key <= NUM_BUCKETS) {
+                if (!nonce_history.nonce_table.contains(nonce_history.next_key)) {
+                    nonce_history.nonce_table.add(
+                        nonce_history.next_key,
+                        empty_bucket()
+                    );
+                };
+                nonce_history.next_key = nonce_history.next_key + 1;
+            }
+        }
+    }
+
+    // Returns true if the input (address, nonce) pair doesn't exist in the nonce history, and inserted into nonce history successfully.
+    // Returns false if the input (address, nonce) pair already exists in the nonce history.
+    public(friend) fun check_and_insert_nonce(
+        sender_address: address,
+        nonce: u64,
+        txn_expiration_time: u64,
+    ): bool acquires NonceHistory {
+        assert!(exists<NonceHistory>(@aptos_framework), error::invalid_state(E_NONCE_HISTORY_DOES_NOT_EXIST));
+        let nonce_history = &mut NonceHistory[@aptos_framework];
+        let nonce_key = NonceKey {
+            sender_address,
+            nonce,
+        };
+        let index = sip_hash_from_value(&nonce_key) % NUM_BUCKETS;
+        let current_time = timestamp::now_seconds();
+        if (!nonce_history.nonce_table.contains(index)) {
+            nonce_history.nonce_table.add(
+                index,
+                empty_bucket()
+            );
+        };
+        let bucket = table::borrow_mut(&mut nonce_history.nonce_table, index);
+        // Garbage collect expired nonces in the bucket.
+        // If the last time a nonce was added to a map is over MAX_EXPIRATION_TIME ago, then the entire map has only expired nonces.
+        // So we can clear the map.
+        for (i in 0..2) {
+            if (bucket.nonce_maps[i].last_stored_time + MAX_EXPIRATION_TIME < current_time) {
+               bucket.nonce_maps[i].nonces.for_each_and_clear(|_, _| {});
```

Comment:
do we have an upper limit of how many items can be there? what if this goes out of gas?

Reply:
I mean I know big ordered map doesn't, but we should limit to avoid GC goes out of gas and all subsequent txns fall into this bucket would fail?

Reply:
At the moment, we are not charging for prologue, so we won't run out of gas?
In the future, when we charge for prologue, we can restrict GC to limit how many entries we clear in one transaction?

Reply:
it's not only about gas, there's hard limit on some pieces like how big the writeset is. in general having unbounded operations especially for system related things is dangerous

Reply:
I believe irrespective of how many changes we make to a storage slot, the entire final value of the slot will be part of the write set. So, if we garbage collect the entire map, the final value (empty values) will be part of the write set. So, it should be okay I guess? 
In case you prefer to garbage collect only finite number of elements per call, I now updated the code to garbage collect at max 30 nonces per call.